firewood/src/merkle.rs

```diff
@@ -1175,6 +1178,220 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
+                                .get_node(child_address)
+                                .map_err(|e| api::Error::InternalError(e.into()))?;
+
+                            let found_key = parents[1..]
+                                .chunks_exact(2)
+                                .map(|parents| (parents[0].1 << 4) + parents[1].1)
+                                .collect::<Vec<u8>>();
+
+                            self.key_state = IteratorState::Iterating {
+                                // continue iterating from here
+                                last_node: current_node,
+                                parents,
+                            };
+
+                            found_key
+                        } else {
+                            // Branch node with no children?
+                            return Poll::Ready(Some(Err(api::Error::InternalError(Box::new(
+                                MerkleError::ParentLeafBranch,
+                            )))));
+                        }
+                    }
+                    NodeType::Leaf(leaf) => {
+                        let mut next = parents.pop();
+                        loop {
+                            match next {
+                                None => return Poll::Ready(None),
+                                Some((parent, child_position)) => {
+                                    // Assume all parents are branch nodes
+                                    let children = parent.inner().as_branch().unwrap().chd();
+
+                                    // we use wrapping_add here because the value might be u8::MAX indicating that
+                                    // we want to go down branch
+                                    let mut child_position = child_position.wrapping_add(1);
+
+                                    if let Some(found_offset) = children[child_position as usize..]
+                                        .iter()
+                                        .position(|&addr| addr.is_some())
+                                    {
+                                        child_position += found_offset as u8;
+                                    } else {
+                                        next = parents.pop();
+                                        continue;
+                                    }
+
+                                    let addr = children[child_position as usize].unwrap();
+
+                                    // we push (node, u8::MAX) because we will add 1, which will wrap to 0
+                                    let child = self
+                                        .merkle
+                                        .get_node(addr)
+                                        .map(|node| (node, u8::MAX))
+                                        .map_err(|e| api::Error::InternalError(e.into()))?;
+
+                                    // TODO: If the branch has a value, then we shouldn't keep_going
```

Comment:
Hmm, would it be better to fix this issue in this PR instead of leaving a TODO? It seems a straightforward fix.

Reply:
Fixed in #346